server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/delivery/MailDispatcher.java

```diff
@@ -0,0 +1,145 @@
+        private MailetContext mailetContext;
+        private Log log;
+
+        public Builder consume(boolean consume) {
+            this.consume = Optional.of(consume);
+            return this;
+        }
+
+        public Builder mailStorer(MailStore mailStore) {
+            this.mailStore = mailStore;
+            return this;
+        }
+
+        public Builder log(Log log) {
+            this.log = log;
+            return this;
+        }
+
+        public Builder mailetContext(MailetContext mailetContext) {
+            this.mailetContext = mailetContext;
+            return this;
+        }
+
+        public MailDispatcher build() throws MessagingException {
+            Preconditions.checkNotNull(mailStore);
+            Preconditions.checkNotNull(log);
+            Preconditions.checkNotNull(mailetContext);
+            return new MailDispatcher(mailStore, consume.or(CONSUME_DEFAULT_VALUE), log, mailetContext);
+        }
+
+    }
+
+    private final MailStore mailStore;
+    private final boolean consume;
+    private final MailetContext mailetContext;
+    private final Log log;
+
+    private MailDispatcher(MailStore mailStore, boolean consume, Log log, MailetContext mailetContext) {
+        this.mailStore = mailStore;
+        this.consume = consume;
+        this.log = log;
+        this.mailetContext = mailetContext;
+    }
+
+    public void dispatch(Mail mail) throws MessagingException {
+        Collection<MailAddress> errors =  customizeHeadersAndDeliver(mail);
+        if (!errors.isEmpty()) {
+            // If there were errors, we redirect the email to the ERROR
+            // processor.
+            // In order for this server to meet the requirements of the SMTP
+            // specification, mails on the ERROR processor must be returned to
+            // the sender. Note that this email doesn't include any details
+            // regarding the details of the failure(s).
+            // In the future we may wish to address this.
+            mailetContext.sendMail(mail.getSender(), errors, mail.getMessage(), Mail.ERROR);
+        }
+        if (consume) {
+            // Consume this message
+            mail.setState(Mail.GHOST);
+        }
+    }
+
+    private Collection<MailAddress> customizeHeadersAndDeliver(Mail mail) throws MessagingException {
+        MimeMessage message = mail.getMessage();
+        // Set Return-Path and remove all other Return-Path headers from the message
+        // This only works because there is a placeholder inserted by MimeMessageWrapper
+        message.setHeader(RFC2822Headers.RETURN_PATH, DeliveryUtils.prettyPrint(mail.getSender()));
+
+        List<String> deliveredToHeader = Arrays.asList(Optional.fromNullable(message.getHeader(DELIVERED_TO)).or(NO_HEADERS));
+        message.removeHeader(DELIVERED_TO);
+
+        Collection<MailAddress> errors = deliver(mail, message);
+
+        for (String deliveredTo : deliveredToHeader) {
```

Comment:
but as messages are copy on write, does it make sense to copy the message before the removal so you don't have to re-add headers to it ?